compiler_gym/datasets/benchmark.py

```diff
@@ -0,0 +1,350 @@
+        """
+        executor = thread_pool.get_thread_pool_executor()
+        futures = (
+            executor.submit(validator, env) for validator in self.validation_callbacks()
+        )
+        for future in as_completed(futures):
+            result: Iterable[ValidationError] = future.result()
+            if result:
+                yield from result
+
+    def validation_callbacks(
+        self,
+    ) -> List[ValidationCallback]:
+        """Return the list of registered validation callbacks.
+
+        :return: A list of callables. See :meth:`add_validation_callback()
+            <compiler_gym.datasets.Benchmark.add_validation_callback>`.
+        """
+        return self._validation_callbacks
+
+    def add_source(self, source: BenchmarkSource) -> None:
+        """Register a new source file for this benchmark.
+
+        :param source: The :class:`BenchmarkSource
+            <compiler_gym.envs.datasets.BenchmarkSource>` to register.
+        """
+        self._sources.append(source)
+
+    def add_validation_callback(
+        self,
+        validation_callback: ValidationCallback,
+    ) -> None:
+        """Register a new validation callback that will be executed on
+        :meth:`validate() <compiler_gym.datasets.Benchmark.validate>`.
+
+        :param validation_callback: A callback that accepts a single
+            :class:`CompilerEnv <compiler_gym.envs.CompilerEnv>` argument and
+            returns an iterable sequence of zero or more :class:`ValidationError
+            <compiler_gym.ValidationError>` tuples. Validation callbacks must be
+            thread safe and must not modify the environment.
+        """
+        self._validation_callbacks.append(validation_callback)
+
+    def write_sources_to_directory(self, directory: Path) -> int:
+        """Write the source files for this benchmark to the given directory.
+
+        This writes each of the :attr:`benchmark.sources
+        <compiler_gym.datasets.Benchmark.sources>` files to disk.
+
+        :param directory: The directory to write results to. If it does not
+            exist, it is created.
+
+        :return: The number of files written.
+        """
+        directory = Path(directory)
+        directory.mkdir(exist_ok=True, parents=True)
+        uniq_paths = set()
+        for filename, contents in self.sources:
+            path = directory / filename
+            uniq_paths.add(path)
+            path.parent.mkdir(exist_ok=True, parents=True)
+            with open(path, "wb") as f:
+                f.write(contents)
+
+        return len(uniq_paths)
+
+    @classmethod
+    def from_file(cls, uri: str, path: Path):
+        """Construct a benchmark from the path to a file.
+
+        :param uri: The URI of the benchmark.
+
+        :param path: A filesystem path.
+
+        :return: A :class:`Benchmark <compiler_gym.datasets.Benchmark>` instance.
+        """
+        return cls(
```

Comment:
What if the path is a dir?

Reply:
Directories are not supported (could be added in the future with a `from_dir()`).

I've added a check that raises `FileNotFoundError` if `path` is not a file.